Windows: Adds scheduled tasks plugin

dgmcdona opened the pull request:
This adds a plugin 'ScheduledTasks' that can decode binary-encoded scheduled tasks from the Windows registry's SOFTWARE hive using a custom reader that extends the `io.BytesIO` class. Decoding operations are intended to be as fault tolerant as possible, swallowing exceptions and returning `None` to account for smear or missing data.

Because each task can have multiple triggers and multiple actions, a single entry is generated for each trigger + action pair. In the event that the either the actions or triggers could not be parsed due to missing or smeared data, an entry will still be generated using the available information from the other registry value, since trigger and action data is stored separately.

Much more information is decoded than is rendered, this was done intentionally to avoid overpopulating the TreeGrid with less pertinent data and to avoid an explosion of trigger/action-specific fields that may not apply to most other entries, while ensuring that the reader is at the correct offset.

Also documents the raised exceptions in the `RegistryHive.get_key()` docstring.

Example task:

```json
{
  "Action": "C:\\Program Files\\Windows Defender\\MpCmdRun.exe",
  "Action Arguments": "Scan -ScheduleJob -ScanTrigger 55 -IdleScheduledJob",
  "Action Context": "LocalSystem",
  "Action Type": "Exe",
  "Creation Time": null,
  "Display Name": null,
  "Enabled": true,
  "Key Name": "{07846B99-B7DC-4599-AEB1-D421B479570F}",
  "Last Run Time": "2023-10-31T22:13:45+00:00",
  "Last Successful Run Time": null,
  "Principal ID": "LocalSystem",
  "Task Name": "Windows Defender Scheduled Scan",
  "Trigger Description": "Run at 2000-01-01T02:36:41+00:00 and repeat every 1 days",
  "Trigger Type": "Time",
  "Working Directory": null,
  "__children": []
}
```

dgmcdona committed:
Windows: Adds scheduled tasks plugin

This adds a plugin 'ScheduledTasks' that can decode binary-encoded
scheduled tasks from the Windows registry's SOFTWARE hive using a custom
reader that extends the `io.BytesIO` class. Decoding operations are
intended to be as fault tolerant as possible, swallowing exceptions and
returning `None` to account for smear or missing data.

Because each task can have mulitple triggers and multiple actions, a
single entry is generated for each trigger + action pair. In the event
that the either the actions could not be parsed or the triggers could
not be parsed due to missing or smeared data, an entry will still be
generated using the available information from the other registry value,
since trigger and action data is stored separately.

Much more information is decoded than is rendered, this was done
intentionally to avoid overpopulating the TreeGrid with less pertinent
data and to avoid an explosion of trigger and action-specific fields that
may not apply to most other entries.

dgmcdona commented on volatility3/framework/plugins/windows/scheduled_tasks.py:
```diff
@@ -0,0 +1,1793 @@
+            if magic is None:
+                break
+
+            ident = reader.read_bstring()
```
fixed

dgmcdona commented on volatility3/framework/plugins/windows/scheduled_tasks.py:
```diff
@@ -0,0 +1,1793 @@
+            return actions
+
+        if version in [2, 3]:
+            action_context = reader.read_bstring()
```
fixed

dgmcdona commented on volatility3/framework/plugins/windows/scheduled_tasks.py:
```diff
@@ -0,0 +1,1793 @@
+
+        version = reader.read_aligned_u1()
+        _start_boundary = reader.read_task_scheduler_time()
+        _end_boundary = reader.read_task_scheduler_time()
```
fixed

dgmcdona commented on volatility3/framework/plugins/windows/scheduled_tasks.py:
```diff
@@ -0,0 +1,1793 @@
+        reader = _ScheduledTasksReader(data)
+
+        version = reader.read_aligned_u1()
+        _start_boundary = reader.read_task_scheduler_time()
```
fixed

dgmcdona commented on volatility3/framework/plugins/windows/scheduled_tasks.py:
```diff
@@ -0,0 +1,1793 @@
+
+        state_name = binascii.hexlify(reader.read(8)).decode("ascii")
+        datalen = reader.read_aligned_u4()
+        _state_data = base64.b64encode(reader.read(datalen))
```
fixed

dgmcdona commented on volatility3/framework/plugins/windows/scheduled_tasks.py:
```diff
@@ -0,0 +1,1793 @@
+
+        repetition_interval_secs = reader.read_u4()
+        _repetition_duration_secs = reader.read_u4()
+        _repetition_duration_secs2 = reader.read_u4()
```
fixed

dgmcdona commented on volatility3/framework/plugins/windows/scheduled_tasks.py:
```diff
@@ -0,0 +1,1793 @@
+        _repetition_duration_secs = reader.read_u4()
+        _repetition_duration_secs2 = reader.read_u4()
+
+        _stop_at_duration_end = reader.read_bool()
```
fixed

dgmcdona commented on volatility3/framework/plugins/windows/scheduled_tasks.py:
```diff
@@ -0,0 +1,1793 @@
+            _trigger_id = reader.read_bstring()
+            reader.seek((8 - (reader.tell() - cur)) % 8, io.SEEK_CUR)  # pad to block
+        else:
+            _trigger_id = None
```
fixed

dgmcdona commented on volatility3/framework/plugins/windows/scheduled_tasks.py:
```diff
@@ -0,0 +1,1793 @@
+
+        if version >= 0x16:
+            cur = reader.tell()
+            _trigger_id = reader.read_bstring()
```
fixed

dgmcdona commented on volatility3/framework/plugins/windows/scheduled_tasks.py:
```diff
@@ -0,0 +1,1793 @@
+            reader.seek((8 - (reader.tell() - cur)) % 8, io.SEEK_CUR)  # pad to block
+
+        if job_schedule is None:
+            return None
```
fixed

dgmcdona commented on volatility3/framework/plugins/windows/scheduled_tasks.py:
```diff
@@ -0,0 +1,1793 @@
+        # print("Cursor position: %d" % reader.tell())
+        if version >= 0x16:
+            cur = reader.tell()
+            _trigger_id = reader.read_bstring()
```
fixed

dgmcdona commented on volatility3/framework/plugins/windows/scheduled_tasks.py:
```diff
@@ -0,0 +1,1793 @@
+        timeout_secs = reader.read_u4()
+
+        repetition_interval_secs = reader.read_u4()
+        _repetition_duration_secs = reader.read_u4()
```
fixed

dgmcdona commented on volatility3/framework/plugins/windows/scheduled_tasks.py:
```diff
@@ -0,0 +1,1793 @@
+        # print("Boundary cursor position: %d" % reader.tell())
+
+        _delay_seconds = reader.read_u4()
+        timeout_secs = reader.read_u4()
```
fixed

dgmcdona commented on volatility3/framework/plugins/windows/scheduled_tasks.py:
```diff
@@ -0,0 +1,1793 @@
+        end_boundary = reader.read_task_scheduler_time()
+        # print("Boundary cursor position: %d" % reader.tell())
+
+        _delay_seconds = reader.read_u4()
```
fixed

dgmcdona commented on volatility3/framework/plugins/windows/scheduled_tasks.py:
```diff
@@ -0,0 +1,1793 @@
+        if not skip_user:
+            skip_sid = reader.read_aligned_u1() != 0
+        else:
+            skip_sid = None
```
fixed

dgmcdona commented on volatility3/framework/plugins/windows/scheduled_tasks.py:
```diff
@@ -0,0 +1,1793 @@
+
+        workdir = reader.read_bstring()
+        if version == 3:
+            _flags = reader.read_u2()
```
fixed

dgmcdona commented on volatility3/framework/plugins/windows/scheduled_tasks.py:
```diff
@@ -0,0 +1,1794 @@
+        )
+
+    @classmethod
+    def parse_actions_value(
+        cls, actions_value: reg_extensions.CM_KEY_VALUE
+    ) -> Optional[ActionSet]:
```
fixed

ikelos commented on volatility3/framework/plugins/windows/scheduled_tasks.py:
```diff
@@ -0,0 +1,1796 @@
+        return is_localized, filetime
+
+    def read_filetime(self) -> Optional[datetime.datetime]:
+        return datetime.datetime.now()
```
This... seems like the wrong thing?  Was this intentional or just a placeholder?

dgmcdona replied:
Yikes! good catch.

ikelos commented on volatility3/framework/renderers/conversion.py:
```diff
@@ -28,6 +28,27 @@ def wintime_to_datetime(
+        format(v, "x")
+        for v in struct.unpack(
+            ">HQ",
+            buf[8:10] + b"\x00\x00" + buf[10:16],
```
Wow... just... wow.  Microsoft do have some interesting people working for them don't they...  5:P

dgmcdona replied:
Yeah, never heard of the u48 type before...

ikelos commented on volatility3/framework/plugins/windows/scheduled_tasks.py:
```diff
@@ -0,0 +1,1796 @@
+        )
+
+
+class TestActionsDecoding(unittest.TestCase):
```
These should be in a separate file.  I'm all for test cases, but not that add huge chunks of data to the end of an already long plugin.  I believe we do have a test directory.  I'd be much happier if all the tests lived in there.  Ideally we'd have a whole pytest setup that unit-tested each of the individual plugins and components of the framework, but I'm not going to put that on you now (and testing what is a giant parsing framework against all possible test cases isn't really feasible).  However, if you could move this code somewhere appropriate with that goal in mind, I'd be very grateful...

dgmcdona replied:
I moved the tests to `test/plugins/windows/test_scheduled_tasks.py`, but i had to do a kind of ugly import hack in `volatility3/framework/symbols/windows/__init__.py` in order to fix a circular import. Not sure right now if there's a better way to resolve that.

ikelos commented on volatility3/framework/plugins/windows/scheduled_tasks.py:
```diff
@@ -0,0 +1,1796 @@
+        mode_index = reader.read_u4()
+        try:
+            if mode_index is not None:
+                mode = list(TimeMode)[mode_index]
```
This feels like we're using the enum the wrong way?  It should contain the value, and then a lookup dictionary should contain the text we want to display?  Converting it to a list makes me worry that the order won't stay consisent, whereas is the enum were:

```
TimeMode.Once = 1
TimeMode.Dailed = 2
...
```

Then it should be possible to just go `TimeMode(mode_index)` and then later look that up in a dictionary if we want the text for it?

ikelos commented on volatility3/framework/plugins/windows/scheduled_tasks.py:
```diff
@@ -0,0 +1,1796 @@
+        data2 = reader.read_u2()
+        data3 = reader.read_u2()
+
+        reader.seek(2, io.SEEK_CUR)  # pad
```
Do you need the `seek_relative` method, if you're always explicitly doing traditional seeks?

dgmcdona replied:
I think I had intended to replace the traditional seeks with this method, but I'll just remove it instead.

ikelos commented on volatility3/framework/plugins/windows/scheduled_tasks.py:
```diff
@@ -0,0 +1,1796 @@
+        filetime = self.decode_filetime()
+        if filetime is None:
+            return None
+        return is_localized, filetime
```
If we know whether it's localized, could we not just return a python datetime object that includes the localization, rather than a filetime?  The user might want the value changed to a timezone of their choosing, so keeping everything in one format we understand seems better than a separate flag (which appears to be ignored later on when this is called)?

dgmcdona replied:
The `is_localized` field is really there for the Windows task scheduler UI to display the 'synchronize across time zones' checkbox and the datetime (converted to local datetime, if the box is unchecked, or as UTC otherwise). In the binary format, the timestamps are all UTC, so this isn't important for determining when something will actually run.

ikelos replied:
Ok, so... do we still need this information or could it be encoded in the timestamp?  Just wonder if it's worth handing back a tuple?

dgmcdona replied:
There's not really anything useful we can do with it now, I'll remove it from the return value.

ikelos reviewed (commented):
Only half reviewed so far, this is a big one, but there's a few comments for you to be getting on with and I'll take another crack at it tomorrow hopefully...

dgmcdona committed:
Windows: Scheduled tasks - code review fixes

ikelos commented on volatility3/framework/symbols/windows/__init__.py:
```diff
@@ -5,11 +5,12 @@
 
 class WindowsKernelIntermedSymbols(intermed.IntermediateSymbolTable):
     def __init__(self, *args, **kwargs) -> None:
+        from volatility3.framework.symbols.windows.extensions import pe, pool, registry
```
This is absolutely not the right way of solving this.  Can you let me know what the problem with the import is that required this move please?

Either the exception traceback or a description of what goes wrong?

dgmcdona replied:
Ha, I had a feeling it was not. This is the original traceback:
```
ImportError: Failed to import test module: test_scheduled_tasks
Traceback (most recent call last):
  File "/Users/dmcdonald/.pyenv/versions/3.8.10/lib/python3.8/unittest/loader.py", line 154, in loadTestsFromName
    module = __import__(module_name)
  File "/Users/dmcdonald/volatility3/test/plugins/windows/test_scheduled_tasks.py", line 6, in <module>
    from volatility3.plugins.windows import scheduled_tasks
  File "/Users/dmcdonald/volatility3/volatility3/framework/plugins/windows/scheduled_tasks.py", line 18, in <module>
    from volatility3.framework.layers import registry
  File "/Users/dmcdonald/volatility3/volatility3/framework/layers/registry.py", line 17, in <module>
    from volatility3.plugins.windows import pslist
  File "/Users/dmcdonald/volatility3/volatility3/framework/plugins/windows/pslist.py", line 14, in <module>
    from volatility3.framework.symbols.windows.extensions import pe
  File "/Users/dmcdonald/volatility3/volatility3/framework/symbols/windows/__init__.py", line 8, in <module>
    from volatility3.framework.symbols.windows.extensions import pe, pool, registry
  File "/Users/dmcdonald/volatility3/volatility3/framework/symbols/windows/extensions/registry.py", line 11, in <module>
    from volatility3.framework.layers.registry import (
ImportError: cannot import name 'RegistryFormatException' from partially initialized module 'volatility3.framework.layers.registry' (most likely due to a circular import) (/Users/dmcdonald/volatility3/volatility3/framework/layers/registry.py)

```

ikelos replied:
Ok, so the cycle is:

layers.registry -> windows.pslist -> symbols.windows...pe -> windows.extensions.{pe,pool,registry} -> layers.registry

Personally, I think the layers.registry should be able to work without pslist, so we need to investigate why that's there.  Core functionality *shouldn't* need a plugin to run...

ikelos replied:
Apparently I added it 3 years ago because windows 10 now has a registry process...  5:S  So I'd almost rather manually implement finding the `Registry` process than pulling in the whole plugin and everything it brings with it, just to get the process list.

Fancy having a go at making that change and then pulling pslist out of the layers.registry code and documenting how it's duplicate code to avoid pulling in complex code?  Either that or we could carve the code out into its own file and import it into both to reduce code duplication?  I'm happy with either, it just walks the ActiveProcessLinks list from PsActiveProcessHead, and then uses that proc.  It can be as cut down as you like, that's literally the only thing it's doing...

dgmcdona committed:
Windows: Scheduled tasks - removes special time type

Removes the TaskSchedulerTime dataclass and discards the
is_localized value parsed with the timestamp.

ikelos commented:
Ok, if you're happy this is working and you didn't have to mess with anything let's merge it.  I'd still like to fix the registry layer pulling in a plugin.  I think I'll file a separate bug to deal with it though...
